src/components/tl/mlx5/tl_mlx5_ib.c

```diff
@@ -0,0 +1,481 @@
+        ibv_close_device(ctx);
+        return UCC_ERR_NO_MESSAGE;
+    }
+    *ib_ctx = ctx;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_mlx5_create_ibv_ctx(char               **ib_devname,
+                                        struct ibv_context **ctx,
+                                        ucc_base_lib_t      *lib)
+{
+    struct ibv_device **dev_list = ibv_get_device_list(NULL);
+    ucc_status_t        status;
+    int                 i;
+
+    if (!dev_list) {
+        tl_debug(lib, "No IB devices are available");
+        return UCC_ERR_NOT_FOUND;
+    }
+
+    if (!(*ib_devname)) {
+        /* If no device was specified by name, loop through all
+           available until first with active port */
+        for (i = 0; dev_list[i]; ++i) {
+            status = create_ctx_for_dev(dev_list[i], ctx);
+            if (UCC_OK == status) {
+                *ib_devname = (char *)ibv_get_device_name(dev_list[i]);
+                status = UCC_OK;
+                goto out;
+            }
+        }
+
+        tl_debug(lib, "No IB devices found");
+        status = UCC_ERR_NOT_FOUND;
+        goto out;
+    } else {
+        for (i = 0; dev_list[i]; ++i) {
+            if (!strcmp(ibv_get_device_name(dev_list[i]), *ib_devname)) {
+                break;
+            }
+        }
+        if (!dev_list[i]) {
+            tl_error(lib, "IB device %s not found", *ib_devname);
+            status = UCC_ERR_NOT_FOUND;
+            goto out;
+        }
+        status = create_ctx_for_dev(dev_list[i], ctx);
+    }
+
+out:
+    if (dev_list) {
+        ibv_free_device_list(dev_list);
+    }
+    return status;
+}
+
+int ucc_tl_mlx5_check_port_active(struct ibv_context *ctx, int port_num)
+{
+    struct ibv_port_attr port_attr;
+
+    ibv_query_port(ctx, port_num, &port_attr);
+    if (port_attr.state == IBV_PORT_ACTIVE &&
+        port_attr.link_layer == IBV_LINK_LAYER_INFINIBAND) {
+        return 1;
+    }
+    return 0;
+}
+
+int ucc_tl_mlx5_get_active_port(struct ibv_context *ctx)
+{
+    struct ibv_device_attr device_attr;
+    int                    i;
+
+    ibv_query_device(ctx, &device_attr);
+    for (i = 1; i <= device_attr.phys_port_cnt; i++) {
+        if (ucc_tl_mlx5_check_port_active(ctx, i)) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+ucc_status_t ucc_tl_mlx5_qp_connect(struct ibv_qp *qp, uint32_t qp_num,
+                                    uint16_t lid, int port, ucc_base_lib_t *lib)
+{
+    int                ret;
+    struct ibv_qp_attr qp_attr;
+
+    memset(&qp_attr, 0, sizeof(qp_attr));
+    qp_attr.qp_state        = IBV_QPS_INIT;
+    qp_attr.pkey_index      = UCC_QP_PKEY_INDEX;
+    qp_attr.port_num        = port;
+    qp_attr.qp_access_flags = UCC_QP_ACCESS_FLAGS;
+    if (ibv_modify_qp(qp, &qp_attr,
+                      IBV_QP_STATE | IBV_QP_PKEY_INDEX | IBV_QP_PORT |
+                          IBV_QP_ACCESS_FLAGS) != 0) {
+        tl_error(lib, "QP RESET->INIT failed");
```

Comment:
Add %m format specifier in error logging for all ibv_xx and mlx5dv_xx APIs errors